docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,616 @@
+passes, e.g. function-level passes, module-level passes, and sequential passes.
+Each subclass itself could act as a pass manager. For instance, they could glob
+the require passes and execute them or build a dependency graph based on the
+given meta data. The full definition of them could be found in
+`src/relay/pass/pass_manager.cc`_
+
+Module-Level Passes
+^^^^^^^^^^^^^^^^^^^
+
+Module level passes are geared mainly for global and inter-procedural
+optimizations (IPO), which is similar to the module pass used in LLVM. Some
+typical passes in Relay that need the global picture of a module, such as to
+a normal form conversion and lambda lifting, etc, fall in this set. At this
+level, users can even add and/or delete functions in a module.
+
+.. code:: c++
+
+    class ModulePassNode : PassNode {
+      PassInfo pass_info;
+      runtime::TypedPackedFunc<Module(Module, PassContext)> pass_func;
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+      // Other members/methods are omitted
+    };
+
+``pass_info`` maintains the information needed by a module-level pass.
+``pass_func`` sketches the real optimization. For example, we may need to
+perform dead code elimination on the module. We could implement the algorithm in
+the ``pass_func`` and let it run on a module. It will then remove the dead code
+including the unused functions in the module. Note that this field is designed
+as a packed function, which enables the implementation of the optimization in
+both C++ and Python.
+
+Function-Level Passes
+^^^^^^^^^^^^^^^^^^^^^
+
+Function-level passes are used to implement various intra function level
+optimizations for a given Relay module. It fetches one function at a time from
+the function list of a module for optimization and yields a rewritten Relay
+function. Most of Relay passes can be classified into this category, such as
+common subexpression elimination and inference simplification, etc.
+
+Note that the scope of passes at this level is a Relay function. Therefore, we
+cannot add or delete a function through these passes as they are not aware of
+the global information.
+
+.. code:: c++
+   
+    class FunctionPassNode : PassNode {
+      PassInfo pass_info;
+      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func;
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+      bool SkipFunction(const Function& func) const;
+      // Other members/methods are omitted...
+    };
+
+``pass_info`` is identical to what we just described in the module pass.
+``pass_func`` takes a function for optimization, it also needs a module as we
+may use it for reporting errors. A function could be annotated with
+"SkipOptimization" so that it will be ignored during optimization.
+
+Sequential Passes
+^^^^^^^^^^^^^^^^^
+
+SequentialPass is similar to Pytorch nn.Sequential that contains a host of
```

Comment:
We probably want `code::` annotations for terms like `SequentialPass` or `nn.Sequential` (just to be consistent)